<!DOCTYPE html>
<html lang="en-us">
<head>

  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="generator" content="Source Themes Academic 2.4.0">
  <meta name="generator" content="Hugo 0.52" />
  <meta name="author" content="Jose Alquicira Hernandez">

  
  
  
  
    
  
  <meta name="description" content="In this tutorial, I’ll show how to incorporate prediction analyses using Seurat along with scPred.
First, we’ll load scPred and the tidyverse:
library(scPred) library(tidyverse) Let’s load Seurat:
library(Seurat) We’ll work with a dataset obtained to analyze the ventral midbrain development in human (see Molecular Diversity of Midbrain Development in Mouse, Human, and Stem Cells). The data will be downloaded from Hemberg’s lab:
download.file(&quot;https://scrnaseq-public-datasets.s3.amazonaws.com/scater-objects/manno_human.rds&quot;, destfile = &quot;~/Downloads/manno_human.rds&quot;) Then we import the data and conver thescater object into Seurat">

  
  <link rel="alternate" hreflang="en-us" href="/tutorial/scpred-integration-with-seurat/">

  


  

  
  
  
  <meta name="theme-color" content="#4caf50">
  

  
  
  
  
    
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.1.3/css/bootstrap.min.css" integrity="sha256-eSi1q2PG6J7g7ib17yAaWMcrr5GrtohYChqibrV7PBE=" crossorigin="anonymous">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/academicons/1.8.6/css/academicons.min.css" integrity="sha256-uFVgMKfistnJAfoCUQigIl+JfUaP47GrRKjf6CTPVmw=" crossorigin="anonymous">
    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.3.1/css/all.css" integrity="sha384-mzrmE5qonljUremFsqc01SB46JvROS7bZs3IO2EmfFsd15uHvIt+Y8vEf7N7fWAU" crossorigin="anonymous">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.css" integrity="sha256-ygkqlh3CYSUri3LhQxzdcm0n1EQvH2Y+U5S2idbLtxs=" crossorigin="anonymous">

    
    
    
      
    
    
      
      
        
          <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" crossorigin="anonymous">
        
      
    

    

    

  

  
  
  <link rel="stylesheet" href=//fonts.googleapis.com/css?family=Montserrat:400,700|Roboto:400,400italic,700|Roboto+Mono>
  

  <link rel="stylesheet" href="/styles.css">
  

  
  
  

  
  <link rel="alternate" href="/index.xml" type="application/rss+xml" title="Jose Alquicira Hernandez">
  <link rel="feed" href="/index.xml" type="application/rss+xml" title="Jose Alquicira Hernandez">
  

  <link rel="manifest" href="/site.webmanifest">
  <link rel="icon" type="image/png" href="/img/icon.png">
  <link rel="apple-touch-icon" type="image/png" href="/img/icon-192.png">

  <link rel="canonical" href="/tutorial/scpred-integration-with-seurat/">

  <meta property="twitter:card" content="summary_large_image">
  
  <meta property="twitter:site" content="@josealquicirah">
  <meta property="twitter:creator" content="@josealquicirah">
  
  <meta property="og:site_name" content="Jose Alquicira Hernandez">
  <meta property="og:url" content="/tutorial/scpred-integration-with-seurat/">
  <meta property="og:title" content="scPred integration with Seurat | Jose Alquicira Hernandez">
  <meta property="og:description" content="In this tutorial, I’ll show how to incorporate prediction analyses using Seurat along with scPred.
First, we’ll load scPred and the tidyverse:
library(scPred) library(tidyverse) Let’s load Seurat:
library(Seurat) We’ll work with a dataset obtained to analyze the ventral midbrain development in human (see Molecular Diversity of Midbrain Development in Mouse, Human, and Stem Cells). The data will be downloaded from Hemberg’s lab:
download.file(&quot;https://scrnaseq-public-datasets.s3.amazonaws.com/scater-objects/manno_human.rds&quot;, destfile = &quot;~/Downloads/manno_human.rds&quot;) Then we import the data and conver thescater object into Seurat">
  
  
    
  <meta property="og:image" content="/img/icon-192.png">
  <meta property="og:locale" content="en-us">
  
  <meta property="article:published_time" content="2019-01-16T00:00:00&#43;00:00">
  
  <meta property="article:modified_time" content="2019-01-16T00:00:00&#43;00:00">
  

  

  

  <title>scPred integration with Seurat | Jose Alquicira Hernandez</title>

</head>
<body id="top" data-spy="scroll" data-target="#TableOfContents" data-offset="71" >

<nav class="navbar navbar-light fixed-top navbar-expand-lg py-0" id="navbar-main">
  <div class="container">

    
      <a class="navbar-brand" href="/">Jose Alquicira Hernandez</a>
      
      <button type="button" class="navbar-toggler" data-toggle="collapse"
              data-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
        <span><i class="fas fa-bars"></i></span>
      </button>
      

    
    <div class="collapse navbar-collapse" id="navbar">

      
      
      <ul class="navbar-nav ml-auto">
        

        

        
        
        
          
        

        <li class="nav-item">
          <a class="nav-link" href="/#about">
            
            <span>Home</span>
            
          </a>
        </li>

        
        

        

        
        
        
          
        

        <li class="nav-item">
          <a class="nav-link" href="/#publications_selected">
            
            <span>Publications</span>
            
          </a>
        </li>

        
        

        

        
        
        
          
        

        <li class="nav-item">
          <a class="nav-link" href="/#posts">
            
            <span>Posts</span>
            
          </a>
        </li>

        
        

        

        
        
        
          
        

        <li class="nav-item">
          <a class="nav-link" href="/tutorial/">
            
            <span>Tutorials</span>
            
          </a>
        </li>

        
        

        

        
        
        
          
        

        <li class="nav-item">
          <a class="nav-link" href="/#contact">
            
            <span>Contact</span>
            
          </a>
        </li>

        
        

      

        

        
      </ul>

    </div>
  </div>
</nav>


<article class="article" itemscope itemtype="http://schema.org/Article">

  


  <div class="article-container">
    <h1 itemprop="name">scPred integration with Seurat</h1>

    

<div class="article-metadata">

  
  
  <span itemscope itemprop="author" itemtype="http://schema.org/Person">
    <meta itemprop="name" content="Jose Alquicira Hernandez">
  </span>
  

  <span class="article-date">
    
    <meta content="2019-01-16 00:00:00 &#43;0000 UTC" itemprop="datePublished">
    <time datetime="2019-01-16 00:00:00 &#43;0000 UTC" itemprop="dateModified">
      Jan 16, 2019
    </time>
  </span>
  <span itemscope itemprop="publisher" itemtype="http://schema.org/Person">
    <meta itemprop="name" content="Jose Alquicira Hernandez">
  </span>

  
  <span class="middot-divider"></span>
  <span class="article-reading-time">
    11 min read
  </span>
  

  
  

  
  
  
  

  
  
<div class="share-box" aria-hidden="true">
  <ul class="share">
    <li>
      <a class="twitter"
         href="https://twitter.com/intent/tweet?text=scPred%20integration%20with%20Seurat&amp;url=%2ftutorial%2fscpred-integration-with-seurat%2f"
         target="_blank" rel="noopener">
        <i class="fab fa-twitter"></i>
      </a>
    </li>
    <li>
      <a class="facebook"
         href="https://www.facebook.com/sharer.php?u=%2ftutorial%2fscpred-integration-with-seurat%2f"
         target="_blank" rel="noopener">
        <i class="fab fa-facebook-f"></i>
      </a>
    </li>
    <li>
      <a class="linkedin"
         href="https://www.linkedin.com/shareArticle?mini=true&amp;url=%2ftutorial%2fscpred-integration-with-seurat%2f&amp;title=scPred%20integration%20with%20Seurat"
         target="_blank" rel="noopener">
        <i class="fab fa-linkedin-in"></i>
      </a>
    </li>
    <li>
      <a class="weibo"
         href="http://service.weibo.com/share/share.php?url=%2ftutorial%2fscpred-integration-with-seurat%2f&amp;title=scPred%20integration%20with%20Seurat"
         target="_blank" rel="noopener">
        <i class="fab fa-weibo"></i>
      </a>
    </li>
    <li>
      <a class="email"
         href="mailto:?subject=scPred%20integration%20with%20Seurat&amp;body=%2ftutorial%2fscpred-integration-with-seurat%2f">
        <i class="fas fa-envelope"></i>
      </a>
    </li>
  </ul>
</div>


  

</div>


    <div class="article-style" itemprop="articleBody">
      


<p>In this tutorial, I’ll show how to incorporate prediction analyses using Seurat along with scPred.</p>
<p>First, we’ll load <code>scPred</code> and the <code>tidyverse</code>:</p>
<pre class="r"><code>library(scPred)
library(tidyverse)</code></pre>
<p>Let’s load <code>Seurat</code>:</p>
<pre class="r"><code>library(Seurat)</code></pre>
<p>We’ll work with a dataset obtained to analyze the ventral midbrain development in human (see <a href="http://dx.doi.org/10.1016/j.cell.2016.09.027"><em>Molecular Diversity of Midbrain Development in Mouse, Human, and Stem Cells</em></a>). The data will be downloaded from <a href="https://hemberg-lab.github.io/scRNA.seq.datasets/human/brain/">Hemberg’s lab</a>:</p>
<pre class="r"><code>download.file(&quot;https://scrnaseq-public-datasets.s3.amazonaws.com/scater-objects/manno_human.rds&quot;, destfile = &quot;~/Downloads/manno_human.rds&quot;)</code></pre>
<p>Then we import the data and conver the<code>scater</code> object into <code>Seurat</code></p>
<pre class="r"><code>midbrain &lt;- readRDS(&quot;~/Downloads/manno_human.rds&quot;)
midbrain &lt;- Convert(from = midbrain, to = &quot;seurat&quot;)</code></pre>
<p>To show the training and prediction steps we’ll split the data into two partitions. First, we generate random indices to select 75% of the total number of cells for training:</p>
<pre class="r"><code>n &lt;- length(midbrain@cell.names)
set.seed(66)
i &lt;- sample(seq_len(n), size = n * 0.75)</code></pre>
<p>Using this indices, we can subset the Seurat object to create two objects containing the training and test data.</p>
<pre class="r"><code>trainData &lt;- SubsetData(midbrain, cells.use = midbrain@cell.names[i])
testData &lt;- SubsetData(midbrain, cells.use = midbrain@cell.names[-i])</code></pre>
<p>We normalize and scale the data using Seurat. The most variable genes are also identified in order to perform dimensionality reduction.</p>
<pre class="r"><code>trainData %&gt;% 
  NormalizeData(normalization.method = &quot;LogNormalize&quot;, 
                scale.factor = 10000) %&gt;% 
    FindVariableGenes(mean.function = ExpMean, 
                      dispersion.function = LogVMR, 
                      x.low.cutoff = 0.0125, 
                      x.high.cutoff = 3, 
                      y.cutoff = 0.5, 
                      do.plot = FALSE) %&gt;% 
    ScaleData() -&gt; trainData</code></pre>
<pre><code>## Scaling data matrix</code></pre>
<div id="principal-component-analysis" class="section level1">
<h1>Principal component analysis</h1>
<p><code>scPred</code> finds a set of discriminative principal components as features to perform the predictions. Using Seurat, we’ll perform a PCA and visualize the results.</p>
<pre class="r"><code>trainData &lt;- RunPCA(trainData, do.print = FALSE)</code></pre>
<p>For this tutorial we’ll focus on predicting the source of the dopaminergic neurons (<strong>ESCs - embryonic stem cells, iPSCs - induced pluripotent stem cells, and ventral midbrain cells</strong>)</p>
<pre class="r"><code>PCAPlot(trainData, group = &quot;Source&quot;)</code></pre>
<p><img src="/tutorial/2019-01-16-scpred-integration-with-seurat_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
</div>
<div id="get-feature-space" class="section level1">
<h1>Get feature space</h1>
<p>The function <code>getFeatureSpace()</code> from <em>scPred</em> can handle Seurat objects as input. The Seurat object must contain a principal component analysis computed already. Here, we provide the <code>trainData</code> object and specify the variable in the metadata slot containing the cell class of interest (e.g. <em>Source</em>).</p>
<pre class="r"><code>scpred &lt;- getFeatureSpace(trainData, pVar = &quot;Source&quot;)</code></pre>
<pre><code>## Not all the classes are valid R variable names</code></pre>
<pre><code>## The following classes are renamed: 
## ventral midbrain</code></pre>
<pre><code>## 
## See new classes in &#39;Source.valid&#39; column in metadata:</code></pre>
<pre><code>## ventral.midbrain</code></pre>
<pre><code>## 
## DONE!</code></pre>
<p>Once we obtain the feature space, the output of the <code>getFeatureSpace()</code> function is a <code>scPred</code> object which contains the normalized expression data, the principal component analysis and the metadata.</p>
</div>
<div id="model-training" class="section level1">
<h1>Model training</h1>
<p>Now we’ll train a support vector machine with a radial kernel using <code>scPred</code> using the <code>trainData</code> function.</p>
<pre class="r"><code>scpred &lt;- trainModel(scpred)</code></pre>
</div>
<div id="prediction" class="section level1">
<h1>Prediction</h1>
<p>The <code>scPredict()</code> function can also handle Seurat objects containing the gene expression data of the cells to be predicted. This functions returns a dataframe stored in the <code>@predictions</code> slot with the class probabilities and the predicted class for each cell in the test dataset.</p>
<pre class="r"><code>scpred &lt;- scPredict(scpred, newData = testData)</code></pre>
<pre><code>## Scaling data matrix</code></pre>
<pre><code>## Predicting cell types</code></pre>
<p>We can get the predictions using the <code>getPredictions()</code> function:</p>
<pre class="r"><code>getPredictions(scpred) %&gt;% head(n = 10)</code></pre>
<pre><code>##                 ventral.midbrain       ESCs        iPSCs  predClass
## 1772122_302_C04     6.017780e-04 0.02568502 0.6178537020 unassigned
## 1772122_302_D11     1.331316e-02 0.52307041 0.0070626037 unassigned
## 1772122_180_A06     6.288517e-04 0.99779341 0.0007134394       ESCs
## 1772122_300_E06     1.833807e-05 0.99998932 0.0001600490       ESCs
## 1772122_300_F10     1.020733e-05 0.99995711 0.0001408679       ESCs
## 1772122_301_C12     2.308561e-03 0.98552436 0.0027696475       ESCs
## 1772122_299_H11     1.973662e-06 0.99998539 0.0002169245       ESCs
## 1772122_301_A02     3.782298e-03 0.99011928 0.0019142387       ESCs
## 1772122_300_D02     1.849093e-05 0.99996094 0.0001692000       ESCs
## 1772122_302_A08     4.622436e-05 0.96686430 0.2695268815       ESCs</code></pre>
<p>In order to get the accuracy of <code>scPred</code>, well add the test metadata to the <code>@predMeta</code> slot. The number of rows in the metadata must match the number of cells that were predicted.</p>
<pre class="r"><code>scpred@predMeta &lt;- testData@meta.data</code></pre>
<p>Once we have the test metadata stored in the <code>scPred</code> object, we can use the <code>crossTab</code> function to get a contingency table showing the predictions versus the true cell classes. In this case, the column <em>Source</em> of the test metadata contains the true cell labels:</p>
<blockquote>
<p>We reorder the rows of the contingency table to match the true cell categories</p>
</blockquote>
<pre class="r"><code>crossTab(scpred, &quot;Source&quot;)[c(1,2,4,3),]</code></pre>
<pre><code>##                  ESCs iPSCs ventral midbrain
## ESCs             0.97  0.03                0
## iPSCs            0.00  0.97                0
## ventral.midbrain 0.00  0.00                1
## unassigned       0.03  0.00                0</code></pre>
<p>Along the diagonal we can see the sensitivity of scPred per class. We can get the same table with counts too:</p>
<pre class="r"><code>crossTab(scpred, &quot;Source&quot;, prop = FALSE)[c(1,2,4,3),]</code></pre>
<pre><code>##                  ESCs iPSCs ventral midbrain
## ESCs              422     3                0
## iPSCs               0    83                0
## ventral.midbrain    0     0              485
## unassigned         15     0                0</code></pre>
</div>
<div id="session-info" class="section level1">
<h1>Session info</h1>
<pre class="r"><code>options(width = 70)
devtools::session_info()</code></pre>
<pre><code>## ─ Session info ─────────────────────────────────────────────────────
##  setting  value                       
##  version  R version 3.5.1 (2018-07-02)
##  os       macOS High Sierra 10.13.6   
##  system   x86_64, darwin15.6.0        
##  ui       X11                         
##  language (EN)                        
##  collate  en_AU.UTF-8                 
##  ctype    en_AU.UTF-8                 
##  tz       Australia/Sydney            
##  date     2019-01-16                  
## 
## ─ Packages ─────────────────────────────────────────────────────────
##  package              * version    date       lib source        
##  abind                  1.4-5      2016-07-21 [1] CRAN (R 3.5.0)
##  acepack                1.4.1      2016-10-29 [1] CRAN (R 3.5.0)
##  ape                    5.2        2018-09-24 [1] CRAN (R 3.5.0)
##  assertthat             0.2.0      2017-04-11 [1] CRAN (R 3.5.0)
##  backports              1.1.2      2017-12-13 [1] CRAN (R 3.5.0)
##  base64enc              0.1-3      2015-07-28 [1] CRAN (R 3.5.0)
##  bibtex                 0.4.2      2017-06-30 [1] CRAN (R 3.5.0)
##  bindr                  0.1.1      2018-03-13 [1] CRAN (R 3.5.0)
##  bindrcpp             * 0.2.2      2018-03-29 [1] CRAN (R 3.5.0)
##  Biobase              * 2.42.0     2018-10-30 [1] Bioconductor  
##  BiocGenerics         * 0.28.0     2018-10-30 [1] Bioconductor  
##  BiocParallel         * 1.16.0     2018-10-30 [1] Bioconductor  
##  bit                    1.1-14     2018-05-29 [1] CRAN (R 3.5.0)
##  bit64                  0.9-7      2017-05-08 [1] CRAN (R 3.5.0)
##  bitops                 1.0-6      2013-08-17 [1] CRAN (R 3.5.0)
##  blogdown               0.9        2018-10-23 [1] CRAN (R 3.5.0)
##  bookdown               0.8        2018-12-03 [1] CRAN (R 3.5.0)
##  broom                  0.5.0      2018-07-17 [1] CRAN (R 3.5.0)
##  callr                  3.0.0      2018-08-24 [1] CRAN (R 3.5.0)
##  caret                * 6.0-80     2018-05-26 [1] CRAN (R 3.5.0)
##  caTools                1.17.1.1   2018-07-20 [1] CRAN (R 3.5.0)
##  cellranger             1.1.0      2016-07-27 [1] CRAN (R 3.5.0)
##  checkmate              1.8.5      2017-10-24 [1] CRAN (R 3.5.0)
##  class                  7.3-14     2015-08-30 [1] CRAN (R 3.5.1)
##  cli                    1.0.1      2018-09-25 [1] CRAN (R 3.5.0)
##  cluster                2.0.7-1    2018-04-13 [1] CRAN (R 3.5.0)
##  codetools              0.2-15     2016-10-05 [1] CRAN (R 3.5.1)
##  colorspace             1.3-2      2016-12-14 [1] CRAN (R 3.5.0)
##  cowplot              * 0.9.3      2018-07-15 [1] CRAN (R 3.5.0)
##  crayon                 1.3.4      2017-09-16 [1] CRAN (R 3.5.0)
##  CVST                   0.2-2      2018-05-26 [1] CRAN (R 3.5.0)
##  data.table             1.11.8     2018-09-30 [1] CRAN (R 3.5.0)
##  ddalpha                1.3.4      2018-06-23 [1] CRAN (R 3.5.0)
##  DelayedArray         * 0.8.0      2018-10-30 [1] Bioconductor  
##  DEoptimR               1.0-8      2016-11-19 [1] CRAN (R 3.5.0)
##  desc                   1.2.0      2018-05-01 [1] CRAN (R 3.5.0)
##  devtools               2.0.1      2018-10-26 [1] CRAN (R 3.5.1)
##  digest                 0.6.18     2018-10-10 [1] CRAN (R 3.5.0)
##  dimRed                 0.1.0      2017-05-04 [1] CRAN (R 3.5.0)
##  diptest                0.75-7     2016-12-05 [1] CRAN (R 3.5.0)
##  doSNOW                 1.0.16     2017-12-13 [1] CRAN (R 3.5.0)
##  dplyr                * 0.7.8      2018-11-10 [1] CRAN (R 3.5.0)
##  DRR                    0.0.3      2018-01-06 [1] CRAN (R 3.5.0)
##  dtw                    1.20-1     2018-05-18 [1] CRAN (R 3.5.0)
##  evaluate               0.12       2018-10-09 [1] CRAN (R 3.5.0)
##  fitdistrplus           1.0-11     2018-09-10 [1] CRAN (R 3.5.0)
##  flexmix                2.3-14     2017-04-28 [1] CRAN (R 3.5.0)
##  forcats              * 0.3.0      2018-02-19 [1] CRAN (R 3.5.0)
##  foreach                1.4.4      2017-12-12 [1] CRAN (R 3.5.0)
##  foreign                0.8-71     2018-07-20 [1] CRAN (R 3.5.0)
##  Formula                1.2-3      2018-05-03 [1] CRAN (R 3.5.0)
##  fpc                    2.1-11.1   2018-07-20 [1] CRAN (R 3.5.0)
##  fs                     1.2.6      2018-08-23 [1] CRAN (R 3.5.0)
##  gbRd                   0.4-11     2012-10-01 [1] CRAN (R 3.5.0)
##  gdata                  2.18.0     2017-06-06 [1] CRAN (R 3.5.0)
##  GenomeInfoDb         * 1.18.1     2018-11-12 [1] Bioconductor  
##  GenomeInfoDbData       1.2.0      2018-11-07 [1] Bioconductor  
##  GenomicRanges        * 1.34.0     2018-10-30 [1] Bioconductor  
##  geometry               0.3-6      2015-09-09 [1] CRAN (R 3.5.0)
##  ggplot2              * 3.1.0      2018-10-25 [1] CRAN (R 3.5.0)
##  ggridges               0.5.1      2018-09-27 [1] CRAN (R 3.5.0)
##  glue                   1.3.0      2018-07-17 [1] CRAN (R 3.5.0)
##  gower                  0.1.2      2017-02-23 [1] CRAN (R 3.5.0)
##  gplots                 3.0.1      2016-03-30 [1] CRAN (R 3.5.0)
##  gridExtra              2.3        2017-09-09 [1] CRAN (R 3.5.0)
##  gtable                 0.2.0      2016-02-26 [1] CRAN (R 3.5.0)
##  gtools                 3.8.1      2018-06-26 [1] CRAN (R 3.5.0)
##  haven                  1.1.2      2018-06-27 [1] CRAN (R 3.5.0)
##  hdf5r                  1.0.1      2018-10-07 [1] CRAN (R 3.5.0)
##  Hmisc                  4.1-1      2018-01-03 [1] CRAN (R 3.5.0)
##  hms                    0.4.2      2018-03-10 [1] CRAN (R 3.5.0)
##  htmlTable              1.12       2018-05-26 [1] CRAN (R 3.5.0)
##  htmltools              0.3.6      2017-04-28 [1] CRAN (R 3.5.0)
##  htmlwidgets            1.3        2018-09-30 [1] CRAN (R 3.5.0)
##  httr                   1.3.1      2017-08-20 [1] CRAN (R 3.5.0)
##  ica                    1.0-2      2018-05-24 [1] CRAN (R 3.5.0)
##  igraph                 1.2.2      2018-07-27 [1] CRAN (R 3.5.0)
##  ipred                  0.9-8      2018-11-05 [1] CRAN (R 3.5.0)
##  IRanges              * 2.16.0     2018-10-30 [1] Bioconductor  
##  irlba                * 2.3.2      2018-01-11 [1] CRAN (R 3.5.0)
##  iterators              1.0.10     2018-07-13 [1] CRAN (R 3.5.0)
##  jsonlite               1.5        2017-06-01 [1] CRAN (R 3.5.0)
##  kernlab              * 0.9-27     2018-08-10 [1] CRAN (R 3.5.0)
##  KernSmooth             2.23-15    2015-06-29 [1] CRAN (R 3.5.1)
##  knitr                  1.20       2018-02-20 [1] CRAN (R 3.5.0)
##  labeling               0.3        2014-08-23 [1] CRAN (R 3.5.0)
##  lars                   1.2        2013-04-24 [1] CRAN (R 3.5.0)
##  lattice              * 0.20-38    2018-11-04 [1] CRAN (R 3.5.0)
##  latticeExtra           0.6-28     2016-02-09 [1] CRAN (R 3.5.0)
##  lava                   1.6.3      2018-08-10 [1] CRAN (R 3.5.0)
##  lazyeval               0.2.1      2017-10-29 [1] CRAN (R 3.5.0)
##  lmtest                 0.9-36     2018-04-04 [1] CRAN (R 3.5.0)
##  lsei                   1.2-0      2017-10-23 [1] CRAN (R 3.5.0)
##  lubridate              1.7.4      2018-04-11 [1] CRAN (R 3.5.0)
##  magic                  1.5-9      2018-09-17 [1] CRAN (R 3.5.0)
##  magrittr               1.5        2014-11-22 [1] CRAN (R 3.5.0)
##  MASS                   7.3-51.1   2018-11-01 [1] CRAN (R 3.5.0)
##  Matrix               * 1.2-15     2018-11-01 [1] CRAN (R 3.5.0)
##  matrixStats          * 0.54.0     2018-07-23 [1] CRAN (R 3.5.0)
##  mclust                 5.4.2      2018-11-17 [1] CRAN (R 3.5.0)
##  memoise                1.1.0      2017-04-21 [1] CRAN (R 3.5.0)
##  metap                  1.0        2018-07-25 [1] CRAN (R 3.5.0)
##  mixtools               1.1.0      2017-03-10 [1] CRAN (R 3.5.0)
##  ModelMetrics           1.2.2      2018-11-03 [1] CRAN (R 3.5.0)
##  modelr                 0.1.2      2018-05-11 [1] CRAN (R 3.5.0)
##  modeltools             0.2-22     2018-07-16 [1] CRAN (R 3.5.0)
##  munsell                0.5.0      2018-06-12 [1] CRAN (R 3.5.0)
##  mvtnorm                1.0-8      2018-05-31 [1] CRAN (R 3.5.0)
##  nlme                   3.1-137    2018-04-07 [1] CRAN (R 3.5.1)
##  nnet                   7.3-12     2016-02-02 [1] CRAN (R 3.5.1)
##  npsurv                 0.4-0      2017-10-14 [1] CRAN (R 3.5.0)
##  pbapply                1.3-4      2018-01-10 [1] CRAN (R 3.5.0)
##  pillar                 1.3.0      2018-07-14 [1] CRAN (R 3.5.0)
##  pkgbuild               1.0.2      2018-10-16 [1] CRAN (R 3.5.1)
##  pkgconfig              2.0.2      2018-08-16 [1] CRAN (R 3.5.0)
##  pkgload                1.0.2      2018-10-29 [1] CRAN (R 3.5.0)
##  pls                    2.7-0      2018-08-21 [1] CRAN (R 3.5.0)
##  plyr                   1.8.4      2016-06-08 [1] CRAN (R 3.5.0)
##  png                    0.1-7      2013-12-03 [1] CRAN (R 3.5.0)
##  prabclus               2.2-6      2015-01-14 [1] CRAN (R 3.5.0)
##  prettyunits            1.0.2      2015-07-13 [1] CRAN (R 3.5.0)
##  processx               3.2.0      2018-08-16 [1] CRAN (R 3.5.0)
##  prodlim                2018.04.18 2018-04-18 [1] CRAN (R 3.5.0)
##  proxy                  0.4-22     2018-04-08 [1] CRAN (R 3.5.0)
##  ps                     1.2.1      2018-11-06 [1] CRAN (R 3.5.0)
##  purrr                * 0.2.5      2018-05-29 [1] CRAN (R 3.5.0)
##  R.methodsS3            1.7.1      2016-02-16 [1] CRAN (R 3.5.0)
##  R.oo                   1.22.0     2018-04-22 [1] CRAN (R 3.5.0)
##  R.utils                2.7.0      2018-08-27 [1] CRAN (R 3.5.0)
##  R6                     2.3.0      2018-10-04 [1] CRAN (R 3.5.0)
##  RANN                   2.6        2018-07-16 [1] CRAN (R 3.5.0)
##  RColorBrewer           1.1-2      2014-12-07 [1] CRAN (R 3.5.0)
##  Rcpp                   1.0.0      2018-11-07 [1] CRAN (R 3.5.0)
##  RcppRoll               0.3.0      2018-06-05 [1] CRAN (R 3.5.0)
##  RCurl                  1.95-4.11  2018-07-15 [1] CRAN (R 3.5.0)
##  Rdpack                 0.10-1     2018-10-04 [1] CRAN (R 3.5.0)
##  readr                * 1.1.1      2017-05-16 [1] CRAN (R 3.5.0)
##  readxl                 1.1.0      2018-04-20 [1] CRAN (R 3.5.0)
##  recipes                0.1.3      2018-06-16 [1] CRAN (R 3.5.0)
##  remotes                2.0.2      2018-10-30 [1] CRAN (R 3.5.1)
##  reshape2               1.4.3      2017-12-11 [1] CRAN (R 3.5.0)
##  reticulate             1.10       2018-08-05 [1] CRAN (R 3.5.0)
##  rlang                  0.3.0.1    2018-10-25 [1] CRAN (R 3.5.0)
##  rmarkdown              1.10       2018-06-11 [1] CRAN (R 3.5.0)
##  robustbase             0.93-3     2018-09-21 [1] CRAN (R 3.5.0)
##  ROCR                   1.0-7      2015-03-26 [1] CRAN (R 3.5.0)
##  rpart                  4.1-13     2018-02-23 [1] CRAN (R 3.5.1)
##  rprojroot              1.3-2      2018-01-03 [1] CRAN (R 3.5.0)
##  rstudioapi             0.8        2018-10-02 [1] CRAN (R 3.5.0)
##  Rtsne                  0.15       2018-11-10 [1] CRAN (R 3.5.0)
##  rvest                  0.3.2      2016-06-17 [1] CRAN (R 3.5.0)
##  S4Vectors            * 0.20.1     2018-11-09 [1] Bioconductor  
##  scales                 1.0.0      2018-08-09 [1] CRAN (R 3.5.0)
##  scPred               * 0.0.0.9000 2018-12-06 [1] local         
##  SDMTools               1.1-221    2014-08-05 [1] CRAN (R 3.5.0)
##  segmented              0.5-3.0    2017-11-30 [1] CRAN (R 3.5.0)
##  sessioninfo            1.1.1      2018-11-05 [1] CRAN (R 3.5.0)
##  Seurat               * 2.3.4      2018-07-17 [1] CRAN (R 3.5.0)
##  sfsmisc                1.1-2      2018-03-05 [1] CRAN (R 3.5.0)
##  SingleCellExperiment * 1.4.0      2018-10-30 [1] Bioconductor  
##  snow                   0.4-3      2018-09-14 [1] CRAN (R 3.5.0)
##  stringi                1.2.4      2018-07-20 [1] CRAN (R 3.5.0)
##  stringr              * 1.3.1      2018-05-10 [1] CRAN (R 3.5.0)
##  SummarizedExperiment * 1.12.0     2018-10-30 [1] Bioconductor  
##  survival               2.43-1     2018-10-29 [1] CRAN (R 3.5.0)
##  tibble               * 1.4.2      2018-01-22 [1] CRAN (R 3.5.0)
##  tidyr                * 0.8.2      2018-10-28 [1] CRAN (R 3.5.0)
##  tidyselect             0.2.5      2018-10-11 [1] CRAN (R 3.5.0)
##  tidyverse            * 1.2.1      2017-11-14 [1] CRAN (R 3.5.0)
##  timeDate               3043.102   2018-02-21 [1] CRAN (R 3.5.0)
##  trimcluster            0.1-2.1    2018-07-20 [1] CRAN (R 3.5.0)
##  tsne                   0.1-3      2016-07-15 [1] CRAN (R 3.5.0)
##  usethis                1.4.0      2018-08-14 [1] CRAN (R 3.5.0)
##  withr                  2.1.2      2018-03-15 [1] CRAN (R 3.5.0)
##  xfun                   0.4        2018-10-23 [1] CRAN (R 3.5.0)
##  xml2                   1.2.0      2018-01-24 [1] CRAN (R 3.5.0)
##  XVector                0.22.0     2018-10-30 [1] Bioconductor  
##  yaml                   2.2.0      2018-07-25 [1] CRAN (R 3.5.0)
##  zlibbioc               1.28.0     2018-10-30 [1] Bioconductor  
##  zoo                    1.8-4      2018-09-19 [1] CRAN (R 3.5.0)
## 
## [1] /Library/Frameworks/R.framework/Versions/3.5/Resources/library</code></pre>
</div>

    </div>

    





    
    

    

    


  </div>
</article>

<div class="container">
  <footer class="site-footer">
  
  <p class="powered-by">
    <a href="/privacy/">Privacy Policy</a>
  </p>
  

  <p class="powered-by">
    &copy; 2018 &middot; 

    Powered by the
    <a href="https://sourcethemes.com/academic/" target="_blank" rel="noopener">Academic theme</a> for
    <a href="https://gohugo.io" target="_blank" rel="noopener">Hugo</a>.

    
    <span class="float-right" aria-hidden="true">
      <a href="#" id="back_to_top">
        <span class="button_icon">
          <i class="fas fa-chevron-up fa-2x"></i>
        </span>
      </a>
    </span>
    
  </p>
</footer>

</div>


<div id="modal" class="modal fade" role="dialog">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <h5 class="modal-title">Cite</h5>
        <button type="button" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="modal-body">
        <pre><code class="tex hljs"></code></pre>
      </div>
      <div class="modal-footer">
        <a class="btn btn-outline-primary my-1 js-copy-cite" href="#" target="_blank">
          <i class="fas fa-copy"></i> Copy
        </a>
        <a class="btn btn-outline-primary my-1 js-download-cite" href="#" target="_blank">
          <i class="fas fa-download"></i> Download
        </a>
        <div id="modal-error"></div>
      </div>
    </div>
  </div>
</div>

    

    
    

    
    
    
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha512-+NqPlbbtM1QqiK8ZAo4Yrj2c4lNQoGv8P79DPtKzj++l5jnN39rHA/xsqn8zE9l0uSoxaCdrOgFs6yjyfbBxSg==" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.3/imagesloaded.pkgd.min.js" integrity="sha512-umsR78NN0D23AzgoZ11K7raBD+R6hqKojyBZs1w8WvYlsI+QuKRGBx3LFCwhatzBunCjDuJpDHwxD13sLMbpRA==" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.1.3/js/bootstrap.min.js" integrity="sha256-VsEqElsCHSGmnmHXGQzvoWjWwoznFSZc6hs7ARLRacQ=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.4/isotope.pkgd.min.js" integrity="sha512-VDBOIlDbuC4VWxGJNmuFRQ0Li0SKkDpmGyuhAG5LTDLd/dJ/S0WMVxriR2Y+CyPL5gzjpN4f/6iqWVBJlht0tQ==" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.js" integrity="sha256-X5PoE3KU5l+JcX+w09p/wHl9AzK333C4hJ2I9S5mD4M=" crossorigin="anonymous"></script>

      
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js" integrity="sha256-/BfiIkHlHoVihZdc6TFuj7MmJ0TWcWsMXkeDFwhi0zw=" crossorigin="anonymous"></script>
        
        <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/r.min.js"></script>
        
        <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/python.min.js"></script>
        
        <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/bash.min.js"></script>
        
      

      
      
    

    <script src="/js/hugo-academic.js"></script>
    

    
    

    
    
    

    
    
    <script>hljs.initHighlightingOnLoad();</script>
    

    
    
    <script>
      const search_index_filename = "/index.json";
      const i18n = {
        'placeholder': "Search...",
        'results': "Search Results",
        'no_results': "No results found"
      };
      const content_type = {
        'post': "Posts",
        'project': "Projects",
        'publication' : "Publications",
        'talk' : "Talks"
        };
    </script>
    

    
    

    
    
    <script id="search-hit-fuse-template" type="text/x-template">
      <div class="search-hit" id="summary-{{key}}">
      <div class="search-hit-content">
        <div class="search-hit-name">
          <a href="{{relpermalink}}">{{title}}</a>
          <div class="article-metadata search-hit-type">{{type}}</div>
          <p class="search-hit-description">{{snippet}}</p>
        </div>
      </div>
      </div>
    </script>
    

    
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/3.2.1/fuse.min.js" integrity="sha256-VzgmKYmhsGNNN4Ph1kMW+BjoYJM2jV5i4IlFoeZA9XI=" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>
    <script src="/js/search.js"></script>
    

    
    

  </body>
</html>

